File: joi/inheritance.py
import sys
input = sys.stdin.readline

class DSU:
    def __init__(self, n):
        self.e = [-1]*n

    def get(self, x):
        if self.e[x] < 0: return x
        self.e[x] = self.get(self.e[x])
        return self.e[x]

    def same_set(self, x, y):
        return self.get(x) == self.get(y)

    def unite(self, x, y):
        x, y = self.get(x), self.get(y)
        if x == y: return False 
        if self.e[x] > self.e[y]: x, y = y, x
        self.e[x] += self.e[y]; self.e[y] = x 
        return True


N, M, K = map(int, input().split())
edges = []
for i in range(M):
    u, v, w = map(int,input().split())
    u -= 1; v -= 1
    edges.append((u,v,w,i))
dsu = [DSU(N) for _ in range(K)] 
edges.sort(key=lambda x: x[2], reverse=True)
ans = [0]*M

for i in range(M):
    e = edges[i]
    lo = 1 
    hi = K+1
    while lo < hi:
        mid = (lo+hi) >> 1
        if mid <= K and dsu[mid-1].same_set(e[0], e[1]):
            lo = mid +1 
        else: hi = mid 
    if lo <= K:
        dsu[lo-1].unite(e[0], e[1])
        ans[e[3]] = lo 
    else: ans[e[3]] = 0
for x in ans:
    print(x) 

File: joi/inheritance.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std; 

template<typename T> using V = vector<T>;
#define vi V<int>

struct DSU {
    vi e; void init(int n) { e = vi(n,-1); }
    int get(int x) { return e[x] < 0 ? x : e[x] = get(e[x]); }
    bool sameSet(int x, int y) { return get(x) == get(y); } 
    bool unite(int x, int y) {
        x = get(x), y = get(y);
        if (x == y) return false; 
        if (e[x] > e[y]) swap(x, y);
        e[x] += e[y], e[y] = x;
        return true;
    }
};

struct edge {
    int u, v, w;
    int id;
}

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr);
    int N, M, K; cin >> N >> M >> K;
    vector<edge> edges(M);
    for (int i = 0; i < M; i++) {
        int u, v, w; cin >> u >> v >> w;
        u--; v--;
        edges[i] = {u,v,w,i};
    }
    sort(edges.begin(), edges.end(), [&](const auto& a, const auto& b) {
        return a.w > b.w;
    });
    vector<DSU> d(K);
    for (auto& dsu : d) {
        dsu.init(N);
    }
    vector<int> ans(M);
    for (int i = 0; i < M; i++) {
        auto& e = edges[i];
        int lo = 1;
        int hi = K+1;
        while (lo < hi) {
            int mid = lo + (hi-lo) / 2;
            if (mid <= K && d[mid-1].sameSet(e.u, e.v)) {
                lo = mid+1;
            } else { hi = mid; }
        }
        if (lo <= K) { ans[e.id]=lo; d[lo-1].unite(e.u,e.v); }
        else { ans[e.id] = 0; }
    }
    for (int x : ans) {
        cout << x << '\n';
    }

    return 0;   
}